### PR TITLE
Dev zhen

### PR DESCRIPTION
**Pull request by Nehzilrz**

### CODE REVIEW

File: src/data/abstract-database.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    abstract ne(val: any): this;
+
+    /** Specifies a path for use with chaining. */
+    abstract where(path?: string | Object, val?: any): this;
+
+    /** Executes the query */
+    abstract exec(callback?: (err: any, res: any) => void): Promise<any>;
+};
+
+class MongoModel implements AbstractModel {
```
Comment (unhealthy) [inline src/data/abstract-database.ts]: Move this class to another file named `MongoModel.ts` or something like that.

File: src/data/abstract-database.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    public select(arg: string | Object): this {
+        mongoose.Query.prototype.select.apply(this.model, arguments);
+        return this;
+    }
+
+    /** Executes the query */
+    public exec(callback?: (err: any, res: any) => void): Promise<any> {
+        mongoose.Query.prototype.exec.apply(this.model, arguments);
```
Comment (unhealthy) [inline src/data/abstract-database.ts]: Try to merge the code in `mongoose-redis-cache.ts` into this function

File: src/data/mongoose-redis-cache.ts [hunk 3]
```diff
@@ -21,8 +21,7 @@ const mongooseRedisCache = function(mongoose, options, callback) {
         });
     }
 
-    mongoose.Query.prototype._exec = mongoose.Query.prototype.exec;
-    mongoose.Query.prototype.exec = (callback) => {
+    mongoose.Query.prototype.fastExec = (callback) => {
```
Comment (unhealthy) [inline src/data/mongoose-redis-cache.ts]: change `(callback)=>{}` to `callback=>`. When there is only one parameter, parentheses is needless.

File: src/data/mongoose-redis-cache.ts [hunk 4]
```diff
@@ -21,8 +21,7 @@ const mongooseRedisCache = function(mongoose, options, callback) {
...
         let self = this;
```
Comment (unhealthy) [inline src/data/mongoose-redis-cache.ts]: If variable will not be re-assigned, use `const` instead of `let`.

File: src/data/abstract-database.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+class MongoDatabase extends AbstractDatabase {
+    constructor(public name: string = 'test') {
+        super(name);
+        this.db = mongoose.createConnection('localhost', name);
+    }
+}*/
+
+
+//abstract database model
+abstract class AbstractModel {
```
Comment (unhealthy) [inline src/data/abstract-database.ts]: `export default abstract class AbstractModel `

File: src/data/abstract-database.ts (hunk 1 above)
Comment (unhealthy) [inline src/data/abstract-database.ts]: `export default abstract class MongoModel `

File: src/data/abstract-database.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        for (var i = 0; i < 100; ++i) {
+            var entity = new this.model({name: "test"+i});
+            entity.save();
+        }
+    }
+    /**
+     * Specifies an $all query condition.
+     */
+    public all(val: number): this {
+        mongoose.Query.prototype.all.apply(this.model, arguments);
```
Comment (unhealthy) [inline src/data/abstract-database.ts]: why not just `return this.model.all(val)`?

Comment (unhealthy) [inline src/data/abstract-database.ts]: the same comment to other function

**Comment by unhealthy** [comment]
I think may be you could define two class, one is `Abstract Database`, another is `Abstract Model`. Both of them use the same API as Mongoose. Then implement two class `Mongo Database`  and `Mongo Model` inherited the two abstract class.
You can have a look at what we have done in the data server, it's quite similar. （你一看就懂...
https://github.com/HKUST-VISLab/vismooc-data-server/blob/dev/mathematician/DB/base_dbhelper.py
https://github.com/HKUST-VISLab/vismooc-data-server/blob/dev/mathematician/DB/mongo_dbhelper.py

**Comment by unhealthy** [comment]
**Coding style needed to pay attention:**
1. use `const` rather than `let`;
2. avoid to call function in the `prototype` explicitly unless necessary.
If you have to call the method in the `prototype`, it's probably that you do not write the code in right way.
For example, 
```typescript
class Model{
    private model;
    public where(path?: string | Object, val?: any): Model{ 
        mongoose.Query.prototype.where.apply( this.query === null ? this.model : this.query, arguments); 
        ...
    }
    ....
}
```
The better way to write this code is separating the Query from Model class, and call the `where` in Query class:
```typescript
class Query{
    private query;
    public where(path?: string | Object, val?: any): Query{ 
        return <Query>this.query.where(path, val); 
        ...
    }
    ....
}
class Model{
    private model;
    // no more where function in Model, you need to use `find()` function to acquire a Query instance
    // public where(path?: string | Object, val?: any): MongoModel { 
    //   mongoose.Query.prototype.where.apply( this.query === null ? this.model : this.query, arguments); 
    //    ...
    // }

    public find():Query{
        return <Query>this,model.find();
    }
    ....
}
```
3.avoid to use `function.apply()` unless necessary. Again, If you have to use `apply` to invoke a function, it's probably that you do not write the code in right way.

**Some features needed to be implemented**:
1. Three abstract class: Database, Model and Query.
2. Three implementation of the three abstract class: MongoDB, MongoModel and MongoQuery.
3. Some basic idea of the structure of database module:
```typescript
import MyDB from './MongoDB';
import MyModel from './MongoModel';
import MyQuery from './MongoQuery';

// connect to a database using the specific address, port and name
const db:MyDB = new MyDB('localhost', '27017', 'atestDB');
// return a collection(model) from a database. I think the schema could be optional
const model:MyModel = db.model("User", [Schema]);
// return a Query from the model instance
const query:MyQuery = model.find();
// query the data
const result = await query.where("name").equal("Brother Li Zhen").asyncExec();
```

File: src/data/abstract-database.ts (hunk 1 above)
Comment (Nehzilrz) [inline src/data/abstract-database.ts]: solved

File: src/data/abstract-database.ts (hunk 2 above)
Comment (Nehzilrz) [inline src/data/abstract-database.ts]: solved

File: src/data/mongoose-redis-cache.ts (hunk 3 above)
Comment (Nehzilrz) [inline src/data/mongoose-redis-cache.ts]: solved

**Pull request merged**